[2.2 regression] error: Invalid index type "slice[int, None, None]" for "str";...

See #9221 (closed)

### Bug Report

Upgrading to mypy 2.3.0 introduced the bug.

Very strangely, reverting to mypy 2.2.0 did *not* change the behavior described next.

Also, the behavior seems independent of which of Leo's GitHub branches is active.

### To Reproduce

```console
cd c:\Repos\leo-editor
python -m leo.scripts.mypy_leo leo
```
The following mypy-leo.cmd script seems to produce the same results:
```cmd
@echo off
cd c:\Repos\leo-editor
python -m leo.scripts.mypy_leo
```

### Actual Behavior

mypy is now complaining about virtually every string slice within Leo. For example:

```python
def findDirective(self, directive_name: str) -> bool:
    if directive_name.startswith('@'):
        directive_name = directive_name[1:]
```
produces this complaint:
```console
leo\core\leoNodes.py:754: error: Invalid index type "slice[int, None, None]" for "str"; expected type "SupportsIndex | slice[SupportsIndex | None, SupportsIndex | None, SupportsIndex]"  [index]
```

### Expected Behavior

No complaints!

### Your Environment

- Mypy version used: mypy 2.3.0 (compiled: yes)
- Mypy command-line flags: None.
  ~~leo.scripts.mypy_leo.py may set flags~~. The results don't  change when using this script.
- Mypy configuration options from `mypy.ini`: See next section
- Python version used: Python 3.14.0

### Summary

This is an extremely strange bug.

Happily, adding `disable_error_code=attr-defined,index` to .mypy.ini calms mypy down.

Any suggestions would be greatly appreciated!

### Latest mypy_leo.py

<details>
<br>

```python

"""
mypy_leo.py: Run mypy on Leo's files.

Info item #3867 describes all of Leo's test scripts:
https://github.com/leo-editor/leo-editor/issues/2867

EKR's mypy-leo.cmd:
    cd {path-to-leo-editor}
    python -m leo.scripts.mypy_leo
"""

# << mypy_leo.py: imports & startup >>
import os
import subprocess
import sys

print(os.path.basename(__file__))

# cd to leo-editor
leo_editor_dir = os.path.abspath(os.path.join(__file__, '..', '..', '..'))
os.chdir(leo_editor_dir)


python = sys.executable
command = rf"{python} -m mypy leo"
subprocess.Popen(command, shell=True).communicate()
```
</details>

### .mypy.ini

<details>
<br>

```ini
# Config file for mypy

# Note: Do not put comments after settings.

[mypy]
python_version = 3.14
cache_dir = mypy_cache
incremental = False

# Defaults for all files.
check_untyped_defs = True
follow_imports = normal
ignore_missing_imports = True
show_error_codes = True
strict_optional = False

# 2026/07/27: These appear to be serious mypy bugs.
disable_error_code=index,arg-type

# For v0.931, per https://github.com/python/mypy/issues/11936

exclude =
    
    # The first line must *not* start with |.
    # Thereafter, each line *must* start with |.
    # No trailing '|' on last entry!

    # Directories...
    doc/|dist/|editpane/|examples/|extensions/|external/|modes/|obsolete/|scripts/|themes/|unittests/|www/|
    
    # Special plugins...
    |freewin.py|scriptFile.py|beautify_node.py|viewrendered3|

    # Experimental plugins...
    |leo_pdf|leo_to_rtf|mnplugins|mod_speedups|settings_finder|stickynotes_plus|
    
    # qt_main.py is generated automatically, so it's pointless to add annotations.
    |qt_main

# More specific descriptions override less specific.

# leo/core...

[mypy-leo.core.*]
disable_error_code=attr-defined
strict_optional = True

[mypy-core.leoPymacs,leo.core.runLeo,leo.core.leoRope]
disable_error_code=attr-defined

[mypy-leo.core.leoclient,leo.leo.core.leoserver]
disable_error_code=attr-defined

# leo/commands.

[mypy-leo.commands.*]
disable_error_code=attr-defined
strict_optional = True

# leo/external.
[mypy-leo.external.*]
disable_error_code=attr-defined

# leo/modes.
[mypy-leo.modes.*]
disable_error_code=attr-defined

# leo/plugins...

[mypy-leo.plugins.*]
disable_error_code=attr-defined

[mypy-leo.plugins.importers.*,leo.plugins.writers.*,leo.unittests.plugins.test_importers]
strict_optional = True

```
</details>

--- a/mypy/typeanal.py
+++ b/mypy/typeanal.py
@@ -2189,7 +2189,7 @@ def fix_instance(
             else:
                 unpacked = [arg]
             for arg in unpacked:
-                with state.strict_optional_set(options.strict_optional):
+                with state.strict_optional_set(True):
                     # Gradually expand defaults, as they may depend on previous variables.
                     if tv.has_default():
                         arg = expand_type(arg, env)

--- a/test-data/unit/check-expressions.test
+++ b/test-data/unit/check-expressions.test
@@ -1248,6 +1248,28 @@ class SupportsIndex(Protocol):
     def __index__(self) -> int: ...
 [builtins fixtures/slice.pyi]
 
+[case testSliceWithPerFileStrictOptional]
+# flags: --no-strict-optional
+# mypy: strict-optional
+"x"[1:]
+[file builtins.py]
+from typing import Generic, TypeVar
+
+_StartT = TypeVar("_StartT", covariant=True)
+_StopT = TypeVar("_StopT", covariant=True, default=_StartT)
+_StepT = TypeVar("_StepT", covariant=True, default=_StartT | _StopT)
+
+class object:
+    def __init__(self) -> None: pass
+class type: pass
+class tuple(Generic[_StartT]): pass
+class function: pass
+class int: pass
+class str:
+    def __getitem__(self, index: slice[int | None]) -> str: pass
+class slice(Generic[_StartT, _StopT, _StepT]): pass
+class dict: pass
+
 [case testNoneSliceBounds]
 from typing import Any
 a: Any

